### DISCUSSION TITLE
Authenticated Telegram User

### DISCUSSION THREAD

**Discussion by iRay** [go-pkgz/auth#157] [Q&A]
Firstly, thank you for a great library!
It is simple and convenient to use.

While implementing oauth for google I needed to get user's email, so I just added custom provider and provided appropriate
scope(userinfo.email) to retrieve user's email.

At the same time, I'm confused about Telegram login.
I created a bot for auth and implemented TelegramHandler. Login through telegram works fine.
I got user's name and avatar.
But they are seems to be useless. It turns out that with telegram login we can only identify that user is a real man and not a bot.
It would be great to have a possibility to send them a message.
In my case: send message to authenticated user when his order is ready.

Would be great to have the ability to fetch auth UserID or ChatID, so we can send him a message.
At the moment user's id is hashed and seems to be useless.
Please, tell me if I want something weird.

Thanks in advance)

**Comment by umputun** [comment]
@paskal - what do you think? Technically it should be easy to fill ChatID into `User.Attributes`; however, I'm unsure if I like it. In fact, it expands the minimal scope and keeps the info auth doesn't need. Maybe an optional parameter for this provider to allow it?

**Comment by paskal** [comment]
@iRay can you please take a look at https://github.com/go-pkgz/notify/blob/master/telegram.go? I don't recall the details, but Telegram notifications work in Remark42 using this module.

**Reply by iRay** [reply]
Yes, I saw that notify module is used.
I see that after telegram auth we're getting user's data from `GetUserInfo` with request and then user's id is get
to send him a notification.

The only thing that I don't understand is that user's id that I get after login is hashed and looks like `telegram_77f9de4fcc0411a247cbdc2ce19a8813ee346830`
that isn't a common telegram user id.

I can't figure out how to get proper id. Maybe I missed something important?
Sorry if the question is dumb.

**Reply by paskal** [reply]
It is constructed as `th.ProviderName + "_" + authtoken.HashID(sha1.New(), fmt.Sprint(update.Message.Chat.ID))` where `update.Message.Chat.ID` is the user ID of the Telegram user. In the current code, getting a real Telegram user ID from the auth module is impossible.

**Reply by iRay** [reply]
Yes, that is why I’m asking about an optional possibility to get telegram user’s id. That is what I described in the first message. And I can’t figure out how messages to users are sent because their ID’s are hashed.

**Reply by paskal** [reply]
It will not be possible to use this library for that: we obfuscate the user credentials by design, and you can't get the user's email address, for example.

**Reply by iRay** [reply]
@paskal Thanks. Understand it. Just asked what if you could add optional param))
Actually we can get email while using google oauth. One need just add "https://www.googleapis.com/auth/userinfo.email" to scopes for handler options.

**Reply by paskal** [reply]
It's possible easily but disabled intentionally: we do not want to store any extra information, and request absolute minimum to identify unique user.
